lib/sycamore/sycamore/transforms/merge_elements.py

```diff
@@ -412,6 +415,136 @@ def merge_elements(self, document: Document) -> Document:
+        else:
+            new_elt.binary_representation = elt1.binary_representation + elt2.binary_representation
+        # Merge text representations by concatenation with a newline
+        if elt1.text_representation is None or elt2.text_representation is None:
+            new_elt.text_representation = elt1.text_representation or elt2.text_representation
+        else:
+            new_elt.text_representation = elt1.text_representation + "\n" + elt2.text_representation
+        # Merge properties by taking the union of the keys
+        properties = new_elt.properties
+        for k, v in elt1.properties.items():
+            properties[k] = v
+            if k == DocumentPropertyTypes.PAGE_NUMBER:
+                properties["page_numbers"] = properties.get("page_numbers", list())
+                properties["page_numbers"] = list(set(properties["page_numbers"] + [v]))
+        for k, v in elt2.properties.items():
+            if properties.get(k) is None:
+                properties[k] = v
+            # if a page number exists, add it to the set of page numbers for this new element
+            if k == DocumentPropertyTypes.PAGE_NUMBER:
+                properties["page_numbers"] = properties.get("page_numbers", list())
+                properties["page_numbers"] = list(set(properties["page_numbers"] + [v]))
+
+        new_elt.properties = properties
+
+        return new_elt
+
+    def customTableHeaderAdditionFilter(self, elements):
+
+        dic = defaultdict(str)
+
+        # First pass: capture headers
+        for ele in elements:
+            if ele.type in ["table", "Image", "Formula"]:
+                continue
+            elif ele.type in ["Text", "Title", "Page-header", "Section-header", "Caption"]:
+                if ele.text_representation is not None:
+                    text_rep = ele.text_representation.strip().lower()
+                if text_rep == "":
+                    continue
+                if re.search(self.regex_pattern, text_rep):
+                    dic[ele["properties"]["page_number"]] = text_rep + " "
+
+        # Second pass: update table elements with headers, done in separate loops since
+        # table headers can be within table elements as well or after them
+        for ele in elements:
+            if ele.type == "table" and isinstance(ele["table"], Table):
+                ele.text_representation = dic[ele["properties"]["page_number"]] + ele.text_representation
+                ele["properties"]["table_header"] = dic[ele["properties"]["page_number"]]
+
+        return elements
+
+    def process_llm_query(self, document):
+        llm_query_agent = LLMTextQueryAgent(prompt=self.llm_prompt, element_type="table", llm=self.llm, table_cont=True)
+        llm_results = llm_query_agent.execute_query(document)
+        return llm_results
+
+    def preprocess_element(self, elem: Element) -> Element:
+        return elem
```

Comment:
I wonder if you can encode enough logic in here and in `should_merge` to not need to override `merge_elements`? prob not I guess since you might need to merge non-adjacent elements?

Reply:
Yes, cause of non-adjacent